### PR TITLE
Feature/api/geo info

### LINKED ISSUE #1003: Colocar as informações dos mapas em rotas diferentes
## Feature Description

Hoje as informações relacionadas ao cadastro de pontos geográficos são colocados na tabela de taxonomia, no entanto, esse formato pode gerar problemas futuros para coletar os dados geográficos mais facilmente.

A sugestão é criar uma nova tabela só para a parte de localização geográfica dentro da plataforma, que poderá ser utilizado pelas diversas features do sistema (tecnologias, organizações, usuários, serviços etc)

Os campos necessários para guardar essa informação são:
- ID da localização 
- Endereço: Descrição do Endereço do local)
- UF (Sigla do Estado com 2 digitos)
- Cidade (com ID de localização do IBGE)
- Latitude (precisão de mais de 7 dígitos após o ponto. Ex: -23.8323034)
- Longitude (precisão de mais de 7 dígitos após o ponto. Ex: -46.1218268)

---------------

## Acceptance criteria

* 

## Implementation Brief

*

### PR DESCRIPTION
**Pull request by alexandreadames**
## Summary

Resolves #1003 

### Rotas novas

| Description | Method | Route | Body/Query | Auth |
| --- | --- | --- | --- | --- |
Lists All Locations | GET | locations | city_id (query param) | NO | 
Gets an single Location | GET | locations/:id| N/A | NO | 
Creates a new location | POST | locations| place_id Mandatory Google Place ID<br>address Mandatory Address<br>city_id Mandatory City ID in cities<br>lat location latitude<br>lng location longitude| YES | 
Updates a location | PUT | locations| place_id Google Place ID<br>address Address<br>city_id Mandatory City ID in cities<br>lat location latitude<br>lng location longitude| YES | 
Deletes a Location | DELETE | locations/:id| N/A | YES | 
Associates location(s) to Technology | POST | technologies/:id/locations | locations: Location ID array in locations | Permissions: UPDATE_TECHNOLOGY or UPDATE_TECHNOLOGIES | 

## QA Steps

Run:
`location.spec.js and technology.spec.js`

## Checklist

- [x] My code is tested and passes existing unit tests.
- [x] My code has an appropriate set of unit tests which all pass.
- [x] My code passes the linting.
- [x] My code has proper inline documentation.
- [x] I have manually tested this PR.

### CODE REVIEW

File: packages/api/database/migrations/1621895452686_location_schema.js [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+		this.create('locations', (table) => {
+			table.increments();
+			table.text('place_id').notNullable();
+			table.text('address').notNullable();
+			table
+				.integer('city_id')
+				.unsigned()
+				.references('id')
+				.inTable('cities')
+				.onDelete('SET NULL');
```
Comment (mateus4k) [inline packages/api/database/migrations/1621895452686_location_schema.js:15]:
```suggestion
				.onUpdate('CASCADE');
				.onDelete('SET NULL');
```

**Review by mateus4k — COMMENTED** [review]
💯

File: packages/api/app/Controllers/Http/TechnologyController.js [hunk 2]
```diff
@@ -285,6 +286,21 @@ class TechnologyController {
...
+		);
+		if (detach) {
+			await technology.locations().detach(null, null, trx);
+		}
+
+		await technology.locations().attach(
+			LocationInstances.map((keyword) => keyword.id),
+			null,
+			trx,
+		);
```
[the remark this answers is not in the document]
Comment (alexandreadames) [reply packages/api/app/Controllers/Http/TechnologyController.js:301]: @mateus acredito que esse código não seja necessário, na verdade ele está redundante, pois no validator eu já checo se o id existe, eu vou retirar.

[the remark this answers is not in the document]
Comment (mateus4k) [reply packages/api/app/Controllers/Http/TechnologyController.js]: Show!

File: packages/api/app/Controllers/Http/TechnologyController.js [hunk 3]
```diff
@@ -285,6 +285,14 @@ class TechnologyController {
 		);
 	}
 
+	async syncronizeLocations(trx, locations, technology, detach = false) {
```
Comment (mateus4k) [inline packages/api/app/Controllers/Http/TechnologyController.js:288]: O parâmetro `detach` está sendo usado em algum canto? Não vi 👀

Comment (alexandreadames) [reply packages/api/app/Controllers/Http/TechnologyController.js:288]: Não, seria para futuras implementações

**Review by mateus4k — APPROVED** [review]

**Pull request merged**